### ISSUE TITLE
Update verl training to support API-model-involving reward on very large batches

### ISSUE THREAD

**Issue by GnarlyMshtep** [GnarlyMshtep/pass_at_k#53]
current state of `pass_at_k` repo is [this](https://github.com/GnarlyMshtep/pass_at_k/commit/39e775156dc8f0921725a1ac141e4477b6721815)

**Why are you implementing the feature? Is it necessary?**
So I can run GnarlyMshtep/cot-monitor#20 and GnarlyMshtep/pass_at_k#64 

**How are you going to implement it? What alternatives have you considered?**
I don't know, I first need a way to test reliability / figure out the problem

### What is happening? What is the behavior I observed? 
To make training more stable, the biggest thing I did was turn up the batch size signficantly. The training script which I started from set it to 1024, and I found that 512 also worked fine. With `n_rollout = 5`, this creates ~ 2500 API queries simultaneously. Doesn't seem like the API handles it well, just hangs for a while -- I tested both openrouter and openai apis. 

When I run 4000 tokens per query via a testing script and 500 tokens each I get reasonable timing 
```
1 Completed 500 tasks in 15.299962759017944 seconds
Starting tasks
2 Completed 500 tasks in 17.559136152267456 seconds
Starting tasks
3 Completed 500 tasks in 15.603731155395508 seconds
Starting tasks
4 Completed 500 tasks in 14.542165756225586 seconds
Starting tasks
5 Completed 500 tasks in 16.256077527999878 seconds
```
(although this is still quite steep and would make for > 1/2 the step meaning that the GPUs would be idle for a long time). 
However when I run the queries within my script, I get 
```
(TaskRunner pid=524087) DEBUG: reward chunking into 6 pieces
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: chunk 0 took 32.12 time
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: chunk 1 took 27.71 time
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: QUERIES TO OA FAILED: Attempt 1 failed, retrying in 1.0 seconds...
(TaskRunner pid=524087) DEBUG: chunk 2 took 74.64 time
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: QUERIES TO OA FAILED: Attempt 1 failed, retrying in 1.0 seconds...
(TaskRunner pid=524087) DEBUG: chunk 3 took 86.07 time
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: chunk 4 took 37.19 time
(TaskRunner pid=524087) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=524087) DEBUG: QUERIES TO OA FAILED: Attempt 1 failed, retrying in 1.0 seconds...
(TaskRunner pid=524087) DEBUG: chunk 5 took 74.31 time
```
which is unacceptably high + sometime freezes. 

### First goal
Iterating on this by editing the verl code and waiting is impractical. I want to try to get a simple out-of-verl repro of this slowness. 

One thing I tried was to repeat a bunch of times in parallel but that did not work, so I can test whether I am improving it / use different APIs. 

I actually have a suspicion that the difference comes from pre-filling: I used the same question in my out-of-verl version -- let me try a variety of questions. I'll use rollouts.

**Comment by GnarlyMshtep** [comment]
The guess is that OA is doing some serious prefill caching which made my first run take 
`1 Completed 1000 tasks in 57.673566818237305 seconds` and future runs take on the order of 16sec. This is a pretty huge difference. It seems this cache is pretty persistent, like over 15min save time. This means that I need to have lots of new data each time to test the API. That... shouldn't be a problem.... I can just read from a huge text file. I can actually generate randomly on the fly (took me like 20min to realize)
```
matanshtepel@GS24461 ai_api_tests % uv run or_limits_test.py
Starting task 1
1 Completed 1000 tasks in 21.139393091201782 seconds
Starting task 2
2 Completed 1000 tasks in 22.662636041641235 seconds
Starting task 2
3 Completed 1000 tasks in 19.83670401573181 seconds
Starting task 3
3 Completed 1000 tasks in 36.0338933467865 seconds
matanshtepel@GS24461 ai_api_tests % uv run oa_limits_test.py
Starting task 1
1 Completed 1000 tasks in 103.85813403129578 seconds
Starting task 2
2 Completed 1000 tasks in 23.97120714187622 seconds
Starting task 2
3 Completed 1000 tasks in 18.264947175979614 seconds
Starting task 3
3 Completed 1000 tasks in 19.313282251358032 seconds
matanshtepel@GS24461 ai_api_tests % 
```
here are some results. The vibe is that somehow openrouter is slightly lower variance? I saw OA spike about 70sec like 3-4 times when testing and never saw it for OR. Will try OR in the code. 

2500 queries (500 toekens out max, in about 7k tokens) consistently freezes, but 1000 consistently does not.

**Comment by GnarlyMshtep** [comment]
```
(TaskRunner pid=69170) DEBUG: reward chunking into 3 pieces
(TaskRunner pid=69170) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=69170) DEBUG: chunk 0 took 11.48 time
(TaskRunner pid=69170) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=69170) DEBUG: chunk 1 took 21.05 time
(TaskRunner pid=69170) DEBUG: starting chunk {chunk_idx}
(TaskRunner pid=69170) DEBUG: chunk 2 took 8.10 time
```
call it a success for now (I think max tokens = 70 also helps which required some code changes)
